api/python/t4/api.py

```diff
@@ -51,13 +51,12 @@ def put(obj, dest, meta=None):
         dest (str): A URI
         meta (dict): Optional. metadata dict to store with ``obj`` at ``dest``
     """
-    data, target = serialize_obj(obj)
-    all_meta = dict(
-        target=target.value,
-        user_meta=meta
-    )
+    all_meta = {'user_meta': meta}
+    clean_dest = fix_url(dest)
+    ext = pathlib.PurePosixPath(urlparse(clean_dest).path).suffix
```

Comment:
Actually, sorry, one more ridiculous URL-related bug: need `unquote(urlparse(clean_dest).path)`.

(E.g., `s3://bucket/foo.%74x%74` is the same URL as `s3://bucket/foo.txt`)

Reply:
Ugh.  yeah, URLs should technically be percent-decoded and then utf-8 decoded -- but then, shouldn't we only accept bytes for URLs, since an encoded URL is a byte-format type?

Reply:
..but no, we're not taking bytes here, we're taking a unicode-encoded object -- and we should use the best format for internal manipulation, and expect that format from our users -- particularly since we're going to be used by hand in (for example) Jupyter notebooks, ipython, bpython, etc.

Reply:
I'm not following this thread. Isn't `unquote` not changing the encoding unless you pass it an optional argument? It's just remapping url escaped characters.  It looks like the the bucket api class we keep s3 urls unescaped internally and we should be consistent about that.

But looking at the bucket api, it looks like these are never escaped when accessing a bucket? Is the s3manager expecting unescaped strings @dimaryaz ?  If so, why ever escape/unescape?

Reply:
Unquote assumes that it's encoded as UTF-8 (which it's supposed to).  When quoting by spec, first the text is encoded as UTF-8, then encoded using percent encoding.  When unquoting, first the bytes are decoded using percent encoding, and then decoded via UTF-8.

The reason this is an issue is that we can't unquote more than once, ultimately.  We need to choose what our "external" or public-facing methods are, and what those accept.  I don't think this is a public-facing method.  If someone uses it, they should know what they're sending.

The only other consistent option is to do it right before needing to use the URL externally -- and that's problematic in our case because we do a lot of internal manipulation of the URLs.

I'd say that right now, we leave this as-is, and later convert everything over to an internal or external URI lib.  In the mean time, we *will* (and already do, in many areas) have inconsistencies with the URL spec.  `file://` when calling `Package.set()` is one example.

Reply:
You can try a path with a space and see what happens, e.g.:
```
urllib.parse.urlparse(fix_url('a b.txt')).path
```
You'll end up with `%20` - so need to `unquote` it.